File: hydra-java/src/test/java/hydra/TestEnv.java
// Hand-written test environment for Java.
// Provides a real graph with primitives and kernel bindings, mirroring Haskell's TestEnv.hs.
// Referenced by the generated TestGraph.java after post-generation patching.

package hydra;

import hydra.graph.Graph;
import hydra.context.Context;

public class TestEnv {
    private static Graph cachedGraph;
    private static Context cachedContext;

    public static synchronized Graph testGraph() {
        if (cachedGraph == null) {
            cachedGraph = TestSuiteRunner.buildTestGraph();
        }
        return cachedGraph;
    }

    public static synchronized Context testContext() {
        if (cachedContext == null) {
            cachedContext = new Context(
                hydra.util.ConsList.empty(),
                hydra.util.ConsList.empty(),
                hydra.util.PersistentMap.empty());
        }
        return cachedContext;
    }
}

File: overlay/scala/hydra-kernel/src/test/scala/hydra/test/testEnv.scala
// Hand-written test environment for Scala.
// Provides a real graph with primitives and kernel bindings, mirroring Haskell's TestEnv.hs.
// Referenced by the generated test data after code generation.

package hydra.test.testEnv

// Signature parity with the DSL: testGraph accepts a Map[Name, Type] and a
// Map[Name, Term] for parity with `testGraph :: Map Name Type -> Map Name Term -> Graph`.
// The actual graph is built from the TestSuiteRunner and ignores both arguments —
// primitives and kernel bindings are host-language specific and can't be expressed
// at the DSL level.
private var cachedGraph: hydra.graph.Graph = null

// Curried signature matches the Scala coder's emission pattern for
// multi-arg DSL functions: Map Name Type -> Map Name Term -> Graph
// becomes (testTypes)(testTerms) at the call site.
//
// The kernel models testGraph polymorphically, so the Scala coder emits the
// call with explicit type arguments — testGraph[Map[Name,Type], Map[Name,Term]]
// — for parity with how it type-applies every generic callee under a type
// application (#434). The two phantom type parameters T0/T1 absorb those args;
// they are unused (the concrete Map parameter types are fixed) but let the
// generated call site type-check.
def testGraph[T0, T1](testTypes: Map[hydra.core.Name, hydra.core.Type])
             (testTerms: Map[hydra.core.Name, hydra.core.Term]): hydra.graph.Graph = {
  if (cachedGraph == null) {
    cachedGraph = hydra.TestSuiteRunner.buildTestGraph()
  }
  cachedGraph
}

lazy val testContext: hydra.typing.InferenceContext = hydra.typing.InferenceContext(
  freshTypeVariableCount = 0,
  trace = Seq.empty)
